xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,379 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/resolver"
+)
+
+var errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+
+func newGracefulSwitchBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *gracefulSwitchBalancer {
+	return &gracefulSwitchBalancer{
+		cc:              cc,
+		bOpts:           opts,
+		scToSubBalancer: make(map[balancer.SubConn]balancer.Balancer),
+		closed:          grpcsync.NewEvent(),
+	}
+}
+
+type gracefulSwitchBalancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu              sync.Mutex
+	balancerCurrent balancer.Balancer
+	balancerPending balancer.Balancer
+	scToSubBalancer map[balancer.SubConn]balancer.Balancer
+	// pendingState is the state of the pending balancer, and is used to update
+	// the Client Conn in the scenario the pending balancer gets swapped to.
+	pendingState     balancer.State
+	currentLBIsReady bool
+
+	// currentMu protects operations on the current balancer.
+	currentMu sync.Mutex
+
+	// closed is fired when this balancer is closed.
```

Comment:
Doug told me to write this :D. I had the same thought as you.

Reply:
I asked for this since there are two child balancers and it was unclear to me when first reading through the code whether it meant our parent closed us or we closed the current/pending balancer.

That said I also prefer short comments on the same line as the field unless a longwinded explanation is actually necessarily, which in this case it is not.